lapce-data/src/hover.rs

```diff
@@ -152,41 +170,317 @@ impl Default for HoverData {
 #[derive(Clone)]
 pub enum HoverItem {
-    // TODO: This could hold the data needed to render the markdown output
-    Markdown(String),
+    /// Rendered (druid) markdown text
+    Markdown(MarkdownText),
     PlainText(String),
 }
 impl HoverItem {
+    pub fn as_markdown_text(&self) -> MarkdownText {
+        match self {
+            HoverItem::Markdown(text) => text.clone(),
+            // TODO: We could store PlainText as an ArcStr/related type to make these clones
+            // significantly cheaper
+            HoverItem::PlainText(text) => MarkdownText::new(
+                RichText::new(ArcStr::from(text.as_str())),
+                Vec::new(),
+            ),
+        }
+    }
+
     pub fn as_str(&self) -> &str {
         match self {
-            HoverItem::Markdown(x) | HoverItem::PlainText(x) => x.as_str(),
+            HoverItem::Markdown(x) => x.as_str(),
+            HoverItem::PlainText(x) => x.as_str(),
         }
     }
-}
-impl From<MarkedString> for HoverItem {
-    fn from(text: MarkedString) -> Self {
+
+    fn from_marked_string(text: MarkedString, style: HoverTextStyle) -> Self {
         match text {
-            MarkedString::String(text) => HoverItem::Markdown(text),
+            MarkedString::String(text) => {
+                HoverItem::Markdown(parse_markdown(&text, style))
+            }
             // This is a short version of a code block
             MarkedString::LanguageString(code) => {
+                // TODO: We could simply construct the MarkdownText directly
                 // Simply construct the string as if it was written directly
-                HoverItem::Markdown(format!(
-                    "```{}\n{}\n```",
-                    code.language, code.value
+                HoverItem::Markdown(parse_markdown(
+                    &format!("```{}\n{}\n```", code.language, code.value),
+                    style,
                 ))
             }
         }
     }
-}
-impl From<MarkupContent> for HoverItem {
-    fn from(content: MarkupContent) -> Self {
+
+    fn from_markup_content(content: MarkupContent, style: HoverTextStyle) -> Self {
         match content.kind {
-            MarkupKind::Markdown => HoverItem::Markdown(content.value),
+            MarkupKind::Markdown => {
+                HoverItem::Markdown(parse_markdown(&content.value, style))
+            }
             MarkupKind::PlainText => HoverItem::PlainText(content.value),
         }
     }
 }
+
+const BLOCKQUOTE_COLOR: Color = Color::grey8(0x88);
+
+#[derive(Data, Clone)]
+pub struct MarkdownText {
+    /// Rendered rich text
+    text: RichText,
+    /// Indices into the underlying text that code blocks start and begin at
+    /// This lets us calculate a rectangle surrounding the text to apply a background too, in order
+    /// to differentiate it from normal text more.
+    /// Note that this covers multiline and inline codeblocks
+    #[data(eq)]
+    code_block_indices: Vec<Range<usize>>,
```

Comment:
Longer term, maybe this list could also store the type of the block. I.e., whether it's a code block, a block quote, or something else. Could be useful if we end up drawing those with different block styles as well.